### ISSUE TITLE
[WIP] Activation Lifecycle [RFC]

### ISSUE THREAD

**Issue by davismj** [aurelia/aurelia#419]
# 💬 RFC

- [ ] [Method Names](#method-names)
- [x] [Execution Order](#execution-order)
- [x] [Arguments](#arguments)
- [ ] Return Values

# The Activation Lifecycle

The activation lifecycle includes a set of custom hooks for developers to leverage, particular for initializing or loading data. The activation lifecycle also allows a developer to deny access to a particular component, and optionally redirect to another component instead.

A very common use case is a component designed for loading a resource. The user will call GET /resource/{id} in `canActivate({ id })`. If a result is found, the data is stored or referenced in the component and `canActivate()` returns true. If no data is found, activation fails and the component isn't loaded.

## Method Names

In vCurrent, the terminology is activate/deactivate. @fkleuver has proposed enter/leave and navigateFrom/To as alternatives (see https://github.com/aurelia/aurelia/issues/293).

The biggest problem with the current naming is that the purpose (described [above](#the-activation-lifecycle)) isn't immediately clear. `activate` suggests "to turn on", but much of the "turning on" behavior is better captured in `attached()` (think event handlers) or `bind()`.

### simple route -> simple route

Let's say you have two routes with single components, `One` and `Two`. Then `can/Deactivate` is called before `can/Activate`.

```typescript
await One.canDeactivate();
await Two.canActivate();
await One.deactivate();
await Two.activate();
```

### hierarchical route -> hierarchical route

Let's say you have two routes with hierarchical components, `OneParent`, `OneChild`, `TwoParent`, and `TwoChild`. Then for `can/Deactivate`, child is called before parent, and for `can/Activate`, parent is called before child.

```typescript
await OneChild.canDeactivate();
await OneParent.canDeactivate();
await TwoParent.canActivate();
await TwoChild.canActivate();
await OneChild.deactivate();
await OneParent.deactivate();
await TwoParent.activate();
await TwoChild.activate();
```

### complex route -> complex route

Let's say you have two routes with independent sibling components, `A`, `B`, `C`, and `D`. Then the sibling components are resolved in parallel for each step.

```typescript
await Promise.all[A.canDeactivate(), B.canDeactivate()];
await Promise.all[C.canActivate(), D.canActivate()];
await Promise.all[A.deactivate(), B.deactivate()];
await Promise.all[C.activate(), D.activate()];
```

### complex hierarchical route -> complex hierarchical route

Let's say you have two routes with hierarchical independent sibling components. Then each hierarchy is run as in the simple case.

```typescript
await Promise.all([
  async () => {
    await AParent.canDeactivate();
    await AChild.canDeactivate();  
  },
  async () => {
    await BParent.canDeactivate();
    await BChild.canDeactivate();  
  }
]);
await Promise.all([
  async () => {
    await CParent.canActivate();
    await CChild.canActivate();
  },
  async () => {
    await DParent.canActivate();
    await DChild.canActivate();
  }
]);
await Promise.all([
  async () => {
    await AParent.deactivate();
    await AChild.deactivate();  
  },
  async () => {
    await BParent.deactivate();
    await BChild.deactivate();  
  }
]);
await Promise.all([
  async () => {
    await CParent.activate();
    await CChild.activate();
  },
  async () => {
    await DParent.activate();
    await DChild.activate();
  }
]);
```

## Arguments

Both the `can/Activate()` and `can/Deactivate()` methods have the same function signature for each lifecycle.

`activate(params, context, navigation)`

**params** `Record<string, any>`

A dictionary of route parameters mapped to their parsed values for the current activation. This provides contextual information for the developer to use when loading data.

**context** `Record<string, any>`

The context for the given activation. This is shared between all methods for a component.

Say you want to check for the existence of resource via HTTP on `canActivate`, and then lock that same resource on `activate`.

```typescript
async canActivate({ id }, context) {
  context.resource = await this.resourceService.get(id);
  return context.resource !== undefined;
}

async activate({ id }, { resource }) {
  this.resource = await this.resourceService.lock(resource);
}
```

**navigation** `NavigationInstruction`

The navigation instruction for the current activation.

**Comment by EisenbergEffect** [comment]
@davismj Do you imaging anything that would plugin and turn the `params` object into something strongly typed? or turn the individual properties into particular types?

**Comment by davismj** [comment]
> or turn the individual properties into particular types?

This. I'm not sure strongly typing the params is worthwhile. Since the primary source of the params is the url, it will always be homomorphic with `Record<string, any>[]`. That means, if you wanted a stronger typing, you could create an interface.

```ts
interface IUserParams {
  userId: int;
}

export class UserRoute {
  activate(params: IUserParams) { }
}
```

**Comment by EisenbergEffect** [comment]
I mean, you could create the interface to get compile-time checking but that wouldn't necessarily match at runtime. So, if it's parsed from the url it will be a string even if you provide a TS interface that says it's an integer. If I recall, there was some proposal to enable typing the parameters as part of the configuration. That would enable the parser to convert things to the correct type.

**Comment by davismj** [comment]
Ok. If I understand you correctly, `params` would not be strongly typed, but the individual properties would based on the route configuration. This proposal is orthogonal. That proposal is here: https://github.com/aurelia/aurelia/issues/164#parameterized-routes

**Comment by EisenbergEffect** [comment]
Groovy. I thought we had talked about that somewhere. Thanks!

**Comment by BBosman** [comment]
Can we use `unknown` where we're proposing `any`? This ensures some runtime type-safety, as the compiler won't let you do stuff without explicitly type-checking (or casting) the params in you code.

**Comment by davismj** [comment]
@BBosman I understand why we want that at an internal level, but at an API level isn't that a bad idea? e.g. my can/Activate almost always looks like this: `activate({ id, name })`. I wouldn't want to get an error here.

Thoughts?

**Comment by fkleuver** [comment]
Yeah for stuff we send out *to* external users we definitely shouldn't use `unknown` as that will force them to cast everything and that's not going to make aurelia very popular. We should however use `unknown` for anything coming *from* users as we cannot make any assumptions

**Comment by BBosman** [comment]
@davismj If `params` is a `Record<string, unknown>` then `activate({ id, name })` is perfectly valid. The question is, what are the types of `id` and `name`?

If you type them as `any` it could be a `string` or a `number` and your IDE/Compiler doesn't know/care. So if you assume it to be a number and use it as such, while it turns out to be a string you'll only notice it at runtime.

If you type them as `unknown` the compiler while complain if you don't verify what it is by a `typeof` check or similar. Or an explicit cast. This will make you write your program in such a way that it's guarded against such runtime errors.

And you still have the option to explicitly type `params` with an interface so you don't need to do type checking. This is essentially an explicit cast as in the above scenario, but for the whole `params` object.

On the other hand, if all params are derived from the url, you could just type it to `Record<string, string>`, but you would have to access it as `params['id']` (or once again create an interface)

So it's mostly a matter of convenience (`any`) or helping users protect themselves (`unknown`).

Personally I'm more in the second camp, but it's not a strong opinion. I do think a discussion on how to handle typing along the edges of our API is important so we can think it through, pick an option and apply it consistently.

**Comment by davismj** [comment]
> If you type them as `any` it could be a `string` or a `number` and your IDE/Compiler doesn't know/care. So if you assume it to be a number and use it as such, while it turns out to be a string you'll only notice it at runtime.

I run into this scenario more often than I'd like to admit.

> So it's mostly a matter of convenience (`any`) or helping users protect themselves (`unknown`).

Even if we did this, it wouldn't bleed through to the activate method because its added to the `Component`, is that right? Or would it be able to somehow?

**Comment by sousekd** [comment]
> `activate(params, context, navigation)`
> 
> **context** `Record<string, any>`
> 
> The context for the given activation. This is shared between all methods for a component.

It would also be nice to have access to **context** mentioned here (or something similar) from children `canActivate` and `activate` handlers. Today, if I want to do functional error handling of navigation failures with child routers, I need to call backend from `canActivate`, and use the loaded data in `activate`. It happens that in order to do an API call in child `canActivate`, I may need the data loaded in parent `canActivate`. I am misusing `navigationInstruction` for this at the moment.

**Comment by Vheissu** [comment]
@bigopon Can this one be closed now?

**Comment by bigopon** [comment]
This is done. We now have activation lifecycle implemented fully, though we will need to have better doc with regards to timing, maybe even a diagram. Though let's chat about it in another issue.
